base/db/R/get.trait.data.R

```diff
@@ -116,144 +104,236 @@ get.trait.data.pft <- function(pft, modeltype, dbfiles, dbcon, trait.names,
-          if (length(dir(pft$outdir, "post.distns.Rdata")) == 0) {
-            all.files <- dir(pft$outdir)
-            post.distn.file <- all.files[grep("post.distns.*.Rdata", all.files)]
+          if (length(list.files(pft$outdir, "post.distns.Rdata")) == 0) {
+            all.files <- list.files(pft$outdir)
+            post.distn.file <- all.files[grep("post\\.distns\\..*\\.Rdata", all.files)]
             if (length(post.distn.file) > 1)
-              stop("get.trait.data.pft() doesn't know how to handle multiple post.distns.*.Rdata files")
+              PEcAn.logger::logger.severe(
+                "get.trait.data.pft() doesn't know how to ",
+                "handle multiple `post.distns.*.Rdata` files.",
+                "Found the following files: ",
+                paste(shQuote(post.distn.file), collapse = ", ")
+              )
             else if (length(post.distn.file) == 1) {
               # Found exactly one post.distns.*.Rdata file. Use it.
-              file.symlink(from = file.path(pft$outdir, post.distn.file),
-                           to = file.path(pft$outdir, 'post.distns.Rdata')
+              link_input <- file.path(pft[["outdir"]], post.distn.file)
+              link_target <- file.path(pft[["outdir"]], "post.distns.Rdata")
+              PEcAn.logger::logger.debug(
+                "Found exactly one posterior distribution file: ",
+                shQuote(link_input),
+                ". Symlinking it to PFT output directory: ",
+                shQuote(link_target)
               )
+              file.symlink(from = link_input, to = link_target)
+            } else {
+              PEcAn.logger::logger.error(
+                "No previous posterior distribution file found. ",
+                "Most likely, trait data were retrieved, but meta-analysis ",
+                "was not run. Meta-analysis will be run."
+              )
+              done <- FALSE
             }
           }
-          return(pft)
+          if (done) return(pft)
         }
       }
     }
   }
 
   # get the trait data (including sampling of derived traits, if any)
-  trait.data <- query.traits(pft_members$id, traits, con = dbcon, update.check.only = FALSE, ids_are_cultivars=(pfttype=="cultivar"))
+  trait.data <- query.traits(pft_members$id, traits, con = dbcon,
+                             update.check.only = FALSE,
+                             ids_are_cultivars = (pfttype == "cultivar"))
   traits <- names(trait.data)
 
   # get list of existing files so they get ignored saving
   old.files <- list.files(path = pft$outdir)
 
   # create a new posterior
   now <- format(x = Sys.time(), format = "%Y-%m-%d %H:%M:%S")
-  db.query(query = paste0("INSERT INTO posteriors (pft_id, created_at, updated_at) VALUES (", pftid, ", '", now, "', '", now, "')"),
+  db.query(paste0("INSERT INTO posteriors (pft_id, created_at, updated_at) ",
```

Comment:
I agree that it's a good idea, but it will require a change in philosophy throughout the entire PEcAn code base, which uses db.query for all database operations. Probably the best thing is to leave db.query as is, but replace our selections with dplyr and write a few custom insert/update/delete functions and start replacing with those little by little.